src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp

```diff
@@ -0,0 +1,200 @@
+#include <Storages/MergeTree/MergeTreeBackgroundExecutor.h>
+
+#include <algorithm>
+
+#include <Common/setThreadName.h>
+#include <Storages/MergeTree/BackgroundJobsAssignee.h>
+
+
+namespace DB
+{
+
+
+String MergeTreeBackgroundExecutor::toString(Type type)
+{
+    switch (type)
+    {
+        case Type::MERGE_MUTATE:
+            return "MergeMutate";
+        case Type::FETCH:
+            return "Fetch";
+        case Type::MOVE:
+            return "Move";
+    }
+    __builtin_unreachable();
+}
+
+
+void MergeTreeBackgroundExecutor::updateConfiguration()
+{
+    auto new_threads_count = std::max<size_t>(1u, threads_count_getter());
+    auto new_max_tasks_count = std::max<size_t>(1, max_task_count_getter());
+
+    try
+    {
+        pending.set_capacity(new_max_tasks_count);
+        active.set_capacity(new_max_tasks_count);
+
+        pool.setMaxFreeThreads(0);
+        pool.setMaxThreads(new_threads_count);
+        pool.setQueueSize(new_max_tasks_count);
+
+        for (size_t number = threads_count; number < new_threads_count; ++number)
+            pool.scheduleOrThrowOnError([this, number] { threadFunction(number); });
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+
+    threads_count = new_threads_count;
+    max_tasks_count = new_max_tasks_count;
+}
+
+
+void MergeTreeBackgroundExecutor::wait()
+{
+    {
+        std::lock_guard lock(mutex);
+        shutdown = true;
+        has_tasks.notify_all();
+    }
+
+    pool.wait();
+}
+
+
+bool MergeTreeBackgroundExecutor::trySchedule(ExecutableTaskPtr task)
+{
+    std::lock_guard lock(mutex);
+
+    if (shutdown)
+        return false;
+
+    try
+    {
+        /// This is needed to increase / decrease the number of threads at runtime
+        if (update_timer.compareAndRestartDeferred(10.))
+            updateConfiguration();
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+
+    auto & value = CurrentMetrics::values[metric];
+    if (value.load() >= static_cast<int64_t>(max_tasks_count))
+        return false;
+
+    pending.push_back(std::make_shared<Item>(std::move(task), metric));
```

Comment:
What is `Item`?  Too generic name for anything.

Reply:
Renamed.